### PR TITLE
Avoid depending on pyqt. Let users choose pyside2

### PR DESCRIPTION
**Pull request by hmaarrfk**
@conda-forge-admin please rerender

Would it be possible not to depend so hard on pyqt5.
I added constraints on pyside2 if that helps assuage fears of integration.

It would be really nice if we could choose between the two without having to download pyqt(5) first.

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -23,10 +23,18 @@ requirements:
   run:
     - python
     - numpy
+  run_constrained:
+    # From the readme
+    # pyqtgraph has had some incompatibilities with PySide2 versions 5.6-5.11, and we recommend you avoid those versions if possible
+    - pyside2 >=5.12
     # pyqtgraph works with pyqt4 or pyqt5. Don't pin it here.
-    - pyqt
+    # - pyqt
```
Comment (marcelotrevisani) [inline recipe/meta.yaml:31]:
```suggestion
     - pyqt
```

Comment (hmaarrfk) [reply recipe/meta.yaml]: it doesn't really constrain it does it :)

Comment (marcelotrevisani) [reply recipe/meta.yaml]: You are right! :)
I just would like to have it there as you are adding pyside there, to remember that the user can chose between those two
:D

**Review by marcelotrevisani — APPROVED** [review]

File: recipe/meta.yaml (hunk 1 above)
Comment (hmaarrfk) [reply recipe/meta.yaml]: yep yep. That is why i had left it there in a comment. Thanks for understanding the need for both!

**Comment by bilderbuchi** [comment]
@hmaarrfk @marcelotrevisani are you sure this patch did the correct thing? AFAICT, with the addition of run_constrained in meta.yml, now *both* pyqt and pyside are considered optional, so you can apparently end up with none installed.

I just installed another package which depends on pyqtgraph in a fresh environment, notice the absence of both pyqt and pyside in the install list:
```
PS C:\> conda install pymeasure
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: C:\Continuum\anaconda3\envs\pymeasure-39

  added / updated specs:
    - pymeasure


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libblas-3.8.0              |           16_mkl         3.6 MB  conda-forge
    libcblas-3.8.0             |           16_mkl         3.6 MB  conda-forge
    liblapack-3.8.0            |           16_mkl         3.6 MB  conda-forge
    numpy-1.19.2               |   py39h5659d19_1         4.9 MB  conda-forge
    pandas-1.1.3               |   py39h6ab05e5_2         9.7 MB  conda-forge
    pymeasure-0.8              |   py39hde42818_2         268 KB  conda-forge
    pyqtgraph-0.11.0           |     pyh9f0ad1d_1         527 KB  conda-forge
    pyvisa-1.10.1              |   py39hde42818_4         216 KB  conda-forge
    ------------------------------------------------------------
                                           Total:        26.2 MB

The following NEW packages will be INSTALLED:

  intel-openmp       pkgs/main/win-64::intel-openmp-2020.1-216
  libblas            conda-forge/win-64::libblas-3.8.0-16_mkl
  libcblas           conda-forge/win-64::libcblas-3.8.0-16_mkl
  liblapack          conda-forge/win-64::liblapack-3.8.0-16_mkl
  mkl                pkgs/main/win-64::mkl-2020.1-216
  numpy              conda-forge/win-64::numpy-1.19.2-py39h5659d19_1
  pandas             conda-forge/win-64::pandas-1.1.3-py39h6ab05e5_2
  pymeasure          conda-forge/win-64::pymeasure-0.8-py39hde42818_2
  pyqtgraph          conda-forge/noarch::pyqtgraph-0.11.0-pyh9f0ad1d_1
  pyserial           pkgs/main/noarch::pyserial-3.4-py_2
  python-dateutil    pkgs/main/noarch::python-dateutil-2.8.1-py_0
  pytz               pkgs/main/noarch::pytz-2020.1-py_0
  pyvisa             conda-forge/win-64::pyvisa-1.10.1-py39hde42818_4
  six                pkgs/main/noarch::six-1.15.0-py_0


Proceed ([y]/n)?
```

And in fact, this leads to a broken pyqtgraph install:
```
PS C:\> python -m pyqtgraph.examples
Traceback (most recent call last):
  File "C:\Continuum\anaconda3\envs\pymeasure-39\lib\runpy.py", line 188, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "C:\Continuum\anaconda3\envs\pymeasure-39\lib\runpy.py", line 111, in _get_module_details
    __import__(pkg_name)
  File "C:\Continuum\anaconda3\envs\pymeasure-39\lib\site-packages\pyqtgraph\__init__.py", line 13, in <module>
    from .Qt import QtGui, mkQApp
  File "C:\Continuum\anaconda3\envs\pymeasure-39\lib\site-packages\pyqtgraph\Qt.py", line 46, in <module>
    raise Exception("PyQtGraph requires one of PyQt4, PyQt5, PySide or PySide2; none of these packages could be imported.")
Exception: PyQtGraph requires one of PyQt4, PyQt5, PySide or PySide2; none of these packages could be imported.
```

Curiously, installing pyqtgraph includes pyqt
```
PS C:\> conda install pyqtgraph
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: C:\Continuum\anaconda3\envs\pymeasure-39

  added / updated specs:
    - pyqtgraph


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    icu-67.1                   |       h33f27b4_0        16.1 MB  conda-forge
    jpeg-9d                    |       he774522_0         344 KB  conda-forge
    libblas-3.8.0              |           16_mkl         3.6 MB  conda-forge
    libcblas-3.8.0             |           16_mkl         3.6 MB  conda-forge
    libclang-10.0.1            |default_hf44288c_1        22.0 MB  conda-forge
    liblapack-3.8.0            |           16_mkl         3.6 MB  conda-forge
    numpy-1.19.2               |   py39h5659d19_1         4.9 MB  conda-forge
    pyqt-5.12.3                |   py39hb0d2dfa_4         4.8 MB  conda-forge
    qt-5.12.9                  |       hb2cf2c5_0       106.3 MB  conda-forge
    ------------------------------------------------------------
                                           Total:       165.2 MB

The following NEW packages will be INSTALLED:

  icu                conda-forge/win-64::icu-67.1-h33f27b4_0
  intel-openmp       pkgs/main/win-64::intel-openmp-2020.1-216
  jpeg               conda-forge/win-64::jpeg-9d-he774522_0
  libblas            conda-forge/win-64::libblas-3.8.0-16_mkl
  libcblas           conda-forge/win-64::libcblas-3.8.0-16_mkl
  libclang           conda-forge/win-64::libclang-10.0.1-default_hf44288c_1
  liblapack          conda-forge/win-64::liblapack-3.8.0-16_mkl
  libpng             pkgs/main/win-64::libpng-1.6.37-h2a8f88b_0
  mkl                pkgs/main/win-64::mkl-2020.1-216
  numpy              conda-forge/win-64::numpy-1.19.2-py39h5659d19_1
  pyqt               conda-forge/win-64::pyqt-5.12.3-py39hb0d2dfa_4
  pyqtgraph          pkgs/main/noarch::pyqtgraph-0.11.0-py_0
  qt                 conda-forge/win-64::qt-5.12.9-hb2cf2c5_0


Proceed ([y]/n)? n
```

**Comment by marcelotrevisani** [comment]
You ended up with none installed because those dependencies are optional so you can install which backend you would like to use
in this case if you want pyside you install it, if you want pyqt you install it

**Comment by bilderbuchi** [comment]
Thank you. I understand that you can choose _which_ backend you use, but one of them has to be present. It seems weird to have to manually install dependencies after installation. 
I have searched a bit and prescribing that one of N (with some default) has to be installed does not seem to be possible.

I am wondering, though: Why does installation work correctly when you install pyqtgraph directly (in this case including pyqt), but not if another package depends on pyqtgraph (and thus pyqt|pyside is a transitive dep)? Is that a problem caused by this PR, or by conda, or something else? Do you see the same behaviour as above?

**Comment by hmaarrfk** [comment]
Personally, I see both pyqt and pyside2 as optional dependencies for this package. 

What they are not, is optional dependencies for your application. If you are concerned that users will not know to install one of them themselves, then you can add one as a dependency to your package.

They have different bug, and different licenses, so I think it is important to leave the freedom of choice at the pyqtgraph level.

Finally, they are both very large packages.

For matplotlib, due to the historic packaging strategy of bundling pyqt, a second package, matplotlib base was added and everything now depends on matplotlib base.

That said, it was a really long transition, that I didnt think was warranted for package, let alone a rather large maintenance burden to keep both packages at infinitum.

**Comment by bilderbuchi** [comment]
> Personally, I see both pyqt and pyside2 as optional dependencies for this package.

Running pyqtgraph (e.g. the examples) does not work (at all?) without one of those "optional" deps present, which does not match my understanding of "optional", but ok, maybe that is a difference of interpretation.

> They have different bug, and different licenses, so I think it is important to leave the freedom of choice at the pyqtgraph level.

absolutely!

**Comment by hmaarrfk** [comment]
> I suspect this is currently not possible with the tooling available.

Correct. My understanding is that it isn't possible to say: at least one of the two libraries.

I think many exceptions in terms of general packaging guidelines are done for qt. I guess this is just an other one of them, and hopefully you understand.

**Comment by bilderbuchi** [comment]
Sure. 
I'm still wondering why `conda install pyqtgraph` chooses pyqt (and the other approach not), though. Is that a bug/feature in conda?

**Comment by hmaarrfk** [comment]
I guess because the anaconda version of the package still depends on pyqt.

https://github.com/AnacondaRecipes/pyqtgraph-feedstock/blob/master/recipe/meta.yaml

It was likely written at a time where PySide2 was a still a far fetched dream and PySide was mostly abbandoned

**Comment by hmaarrfk** [comment]
I that note, i strongly discourage adding 

```
-c conda-forge
```
to the command line.

It typically leads to unstable environments.

Rather, I would suggest you permenantly add conda-forge to the list of channels you subscribe to.

**Comment by bilderbuchi** [comment]
Ha, that could be it! 
I did not add the flag, we have a channel configuration that appends conda-forge after main, so we should primarily get the main packages, and conda-forge is used for what is not available from main.

**Comment by ksunden** [comment]
I will also note that upstream we do not specify either pyside or pyqt as a hard dependency (in fact, numpy is the _only_ dependency specified in our `setup.py` (of course, one of the qt bindings is needed to actually run, but there aren't really good mechanisms in either pypi or conda for saying "install one of these two", as there are in some package mangers)

**Comment by hmaarrfk** [comment]
> so we should primarily get the main packages, and conda-forge is used for what is not available from main.

conda-forge is often ahead of main in certain aspects. I recommend doing it the other way around.

**Comment by bilderbuchi** [comment]
> conda-forge is often ahead of main in certain aspects. I recommend doing it the other way around.

yeah I know, thanks. This approach came out of preferring the more vetted/stabler/less churning "mainline" packages, while at the same time enabling the folks in my lab to not have to use `pip` _at all_, e.g. if they want to install packages that are not on main. So far this has worked out very well.

**Comment by hmaarrfk** [comment]
Hmm. Interesting good to know actually.

**Pull request merged**